File: packages/php/src/Resources/Contacts.php
<?php

declare(strict_types=1);

namespace Axene\Mailer\Resources;

use Axene\Mailer\Internal;
use Axene\Mailer\Transport;

/**
 * The `contacts` resource: manage subscriber lists, their contacts, CSV
 * imports, and templated bulk sends. Accessed as `$client->contacts`.
 */
final class Contacts
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * List all subscriber lists in the active workspace.
     *
     * @return list<array<string, mixed>>
     */
    public function listLists(): array
    {
        return $this->transport->request('GET', '/v1/contacts/');
    }

    /**
     * Create a subscriber list.
     *
     * @param array{name: string, description?: string|null, iconSeed?: string|null} $params
     * @return array<string, mixed>
     */
    public function createList(array $params): array
    {
        return $this->transport->request('POST', '/v1/contacts/', Internal::prune([
            'name' => $params['name'] ?? null,
            'description' => $params['description'] ?? null,
            'icon_seed' => $params['iconSeed'] ?? null,
        ]));
    }

    /**
     * Get a list with a page of its contacts (zero-based `page`).
     *
     * @param array{page?: int, limit?: int} $params
     * @return array<string, mixed>
     */
    public function getList(string $id, array $params = []): array
    {
        return $this->transport->request('GET', '/v1/contacts/' . rawurlencode($id) . Internal::query($params));
    }

    /**
     * Update a list's name, description, or icon (partial).
     *
     * @param array{name?: string, description?: string|null, iconSeed?: string|null} $params
     * @return array<string, mixed>
     */
    public function updateList(string $id, array $params): array
    {
        return $this->transport->request('PATCH', '/v1/contacts/' . rawurlencode($id), Internal::prune([
            'name' => $params['name'] ?? null,
            'description' => $params['description'] ?? null,
            'icon_seed' => $params['iconSeed'] ?? null,
        ]));
    }

    /**
     * Delete a list and all of its contacts.
     */
    public function deleteList(string $id): void
    {
        $this->transport->request('DELETE', '/v1/contacts/' . rawurlencode($id));
    }

    /**
     * Add a single contact to a list.
     *
     * @param array{email: string, name?: string|null, metadata?: array<string, mixed>|null} $params
     * @return array<string, mixed>
     */
    public function addContact(string $listId, array $params): array
    {
        return $this->transport->request(
            'POST',
            '/v1/contacts/' . rawurlencode($listId) . '/contacts',
            Internal::prune([
                'email' => $params['email'] ?? null,
                'name' => $params['name'] ?? null,
                'metadata' => $params['metadata'] ?? null,
            ]),
        );
    }

    /**
     * Remove a contact from a list.
     */
    public function removeContact(string $listId, string $contactId): void
    {
        $this->transport->request(
            'DELETE',
            '/v1/contacts/' . rawurlencode($listId) . '/contacts/' . rawurlencode($contactId),
        );
    }

    /**
     * Import contacts from a CSV file (header row required). Sends the file as
     * multipart/form-data under the field name `file`.
     *
     * @param string $fileContents Raw bytes of the CSV file.
     * @return array<string, mixed>
     */
    public function uploadCsv(string $listId, string $fileContents, string $filename = 'contacts.csv'): array
    {
        return $this->transport->upload(
            '/v1/contacts/' . rawurlencode($listId) . '/upload',
            $fileContents,
            $filename,
        );
    }

    /**
     * Send a templated email to every contact in a list. The list id is
     * injected as `contact_list_id` automatically. Subject/html/text may use
     * `{{email}}`, `{{name}}`, and `{{metadata_key}}` placeholders.
     *
     * @param array{senderAddressId: string, subject: string, html?: string|null, text?: string|null, tags?: list<string>|null} $params
     * @return array<string, mixed>
     */
    public function bulkSend(string $listId, array $params): array
    {
        return $this->transport->request(
            'POST',
            '/v1/contacts/' . rawurlencode($listId) . '/send',
            Internal::prune([
                'contact_list_id' => $listId,
                'sender_address_id' => $params['senderAddressId'] ?? null,
                'subject' => $params['subject'] ?? null,
                'html' => $params['html'] ?? null,
                'text' => $params['text'] ?? null,
                'tags' => $params['tags'] ?? null,
            ]),
        );
    }
}

File: packages/rust/src/resources/contacts.rs
//! The `contacts` resource: subscriber lists, contacts, CSV imports, bulk sends.

use reqwest::Method;

use super::{query, urlencode};
use crate::error::Result;
use crate::http::HttpTransport;
use crate::models::{
    AddContact, BulkSend, BulkSendResult, Contact, ContactList, ContactListDetail, CreateList,
    CsvImportResult, UpdateList,
};

/// Accessed as `client.contacts()`.
#[derive(Debug, Clone)]
pub struct Contacts {
    http: HttpTransport,
}

impl Contacts {
    pub(crate) fn new(http: HttpTransport) -> Self {
        Self { http }
    }

    /// List all subscriber lists in the active workspace.
    pub async fn list_lists(&self) -> Result<Vec<ContactList>> {
        self.http.request(Method::GET, "/v1/contacts/").await
    }

    /// Create a subscriber list.
    pub async fn create_list(&self, params: &CreateList) -> Result<ContactList> {
        self.http
            .request_json(Method::POST, "/v1/contacts/", params)
            .await
    }

    /// Get a list with a page of its contacts (zero-based `page`).
    pub async fn get_list(
        &self,
        id: &str,
        page: Option<u64>,
        limit: Option<u64>,
    ) -> Result<ContactListDetail> {
        let q = query(&[
            ("page", page.map(|p| p.to_string())),
            ("limit", limit.map(|l| l.to_string())),
        ]);
        self.http
            .request(Method::GET, &format!("/v1/contacts/{}{q}", urlencode(id)))
            .await
    }

    /// Update a list's name, description, or icon (partial).
    pub async fn update_list(&self, id: &str, params: &UpdateList) -> Result<ContactList> {
        self.http
            .request_json(
                Method::PATCH,
                &format!("/v1/contacts/{}", urlencode(id)),
                params,
            )
            .await
    }

    /// Delete a list and all of its contacts.
    pub async fn delete_list(&self, id: &str) -> Result<()> {
        self.http
            .request_empty(Method::DELETE, &format!("/v1/contacts/{}", urlencode(id)))
            .await
    }

    /// Add a single contact to a list.
    pub async fn add_contact(&self, list_id: &str, params: &AddContact) -> Result<Contact> {
        self.http
            .request_json(
                Method::POST,
                &format!("/v1/contacts/{}/contacts", urlencode(list_id)),
                params,
            )
            .await
    }

    /// Remove a contact from a list.
    pub async fn remove_contact(&self, list_id: &str, contact_id: &str) -> Result<()> {
        self.http
            .request_empty(
                Method::DELETE,
                &format!(
                    "/v1/contacts/{}/contacts/{}",
                    urlencode(list_id),
                    urlencode(contact_id)
                ),
            )
            .await
    }

    /// Import contacts from a CSV file (header row required). Sent as
    /// `multipart/form-data` under the field `file`.
    pub async fn upload_csv(
        &self,
        list_id: &str,
        file: Vec<u8>,
        filename: &str,
    ) -> Result<CsvImportResult> {
        self.http
            .upload(
                &format!("/v1/contacts/{}/upload", urlencode(list_id)),
                file,
                filename,
            )
            .await
    }

    /// Send a templated email to every contact in a list. The `contact_list_id`
    /// field is set automatically to `list_id`.
    pub async fn bulk_send(&self, list_id: &str, params: &BulkSend) -> Result<BulkSendResult> {
        let mut body = params.clone();
        body.contact_list_id = list_id.to_string();
        self.http
            .request_json(
                Method::POST,
                &format!("/v1/contacts/{}/send", urlencode(list_id)),
                &body,
            )
            .await
    }
}
